### PR TITLE
Add support for enforcer implementations to pick their own admin client implementations

### PR DESCRIPTION
**Pull request by danielfritsch**
Upcoming quota enforcement implementation will require using a Zookeeper admin client in order to add/delete configured quotas. This commit removes the strict binding of the EnforcerCommand with kafka-based AdminClient. 
Instead, we allow the Enforcer implementations to choose and initialize whichever admin clients they need.

### CODE REVIEW

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java [hunk 1]
```diff
@@ -79,13 +85,35 @@ protected EnforceCommand() {
         "for a test.");
   }
 
+  /**
+   * Initialize the enforcer with a zookeeper based admin client. Base classes must override this method.
```
Comment (jyates) [inline kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java:89]: I wouldn't expect all Enforcer impls to use ZK, so wouldn't expect this requirement.

In general, it would be nice if we could find a way to instead have the enforcer impl "pick" the kind of client it needs (could easily see one that needs both Kafka and ZK) and give it access to standard providers for the admin client it needs

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java [hunk 2]
```diff
@@ -53,6 +57,8 @@
       arity = 1)
   protected int interval = 600;
 
+  private static final int TIMEOUT_MS = 30000;
```
Comment (jyates) [inline kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java:60]: looks like this is only used for ZK, a more specific name and more information on what is the timeout for would be helpful

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/BaseCommand.java [hunk 3]
```diff
@@ -81,6 +82,10 @@ public BaseCommand(Map<String, Object> cmdConfig, String cluster) {
     return MAPPER.convertValue(cmdConfig().get("kafka"), MAP_TYPE);
   }
 
+  public String getZookeeperConnection() {
```
Comment (jyates) [inline kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/BaseCommand.java:85]: I imagine there are a number of ZK configuration properties besides just the connection string that we might want - timeouts look to be one of them. In generally, this property is the `zookeeper.connect` string. 

So a parent `zookeeper` section in the YAML would also be appropriate and we could have a similar model to the kafkaConfig block above

**Review by jyates — COMMENTED** [review]
In general, i think we can abstract out the admin logic into a generic provider-like interface (and associated impls) that can then be used in various places where you need access. It looks like we need ways to:
 * specify the type of provider
 * check the base config of the provider
 * load the admin impl from the provider

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/BaseCommand.java (hunk 3 above)
Comment (danielfritsch) [reply kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/BaseCommand.java]: makes sense. Removed this bit for now as the refactored code doesn't explicitly deal with Zooekeeper. Will keep this in mind for upcoming PRs though

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java (hunk 2 above)
Comment (danielfritsch) [reply kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java]: removed ZK bits from this PR as not relevant to current changes, will keep this in mind

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java (hunk 1 above)
Comment (danielfritsch) [reply kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/EnforceCommand.java]: addressed this by using ServiceLoaders: https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html

Enforcer implementations can now pick which of the available client providers to use, and can load the client directly from the provider

**Comment by danielfritsch** [comment]
Updates
- refactored to use [ServiceLoaders](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html), allowing enforcer impls to pick which client(s) to use
- removed any ZK related changes that were present in previous commit, these will come in following PRs

File: kafka_enforcer_common/src/main/resources/META-INF/services/com.tesla.data.enforcer.ClientProvider [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# list of all client providers available to be used
+com.tesla.data.enforcer.KafkaAdminClientProvider
```
Comment (jyates) [inline kafka_enforcer_common/src/main/resources/META-INF/services/com.tesla.data.enforcer.ClientProvider:2]: I'd prefer we stay away from service loader stuff like this. We are only going to have a limited number of impls and no need to have users dynamically specify the services they want to load.

Essentially this is way overkill.

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java [hunk 5]
```diff
@@ -14,17 +15,24 @@
...
 
   private static class TopicEnforceCommand extends EnforceCommand<ConfiguredTopic> {
     @Override
-    protected Enforcer<ConfiguredTopic> initEnforcer(AdminClient adminClient) {
-      return new TopicEnforcer(
-          new TopicServiceImpl(adminClient, dryrun),
-          configuredEntities(ConfiguredTopic.class, "topics", "topicsFile"),
-          !unsafemode);
+    protected Enforcer<ConfiguredTopic> initEnforcer(ServiceLoader<ClientProvider> clientLoader) {
+      // Iterate through all available client providers, and pick the one for kafka AdminClient
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java:26]: instead, it can just specify in the constructor that it wanted the Kafka type admin client provider

File: kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/KafkaAdminClientProvider.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package com.tesla.data.enforcer;
+
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.KafkaAdminClient;
+
+import java.util.Map;
+
+public class KafkaAdminClientProvider implements ClientProvider {
```
Comment (jyates) [inline kafka_enforcer_common/src/main/java/com/tesla/data/enforcer/KafkaAdminClientProvider.java:8]: I would expect the Kafka one to also provide a method to get the config block given a `Map<String, Object>` command config... or at least a config key.

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java (hunk 5 above)
Comment (danielfritsch) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java:26]: I think the issue here is that initEnforcer is called from EnforceCommand (not from the Acl/Topic EnforceCommand impls, which know what kind of client it needs), and so you can't just pass in which type of client it needs like that.

I think a reasonable workaround can be to just have initEnforcer take no arguments, and let the individual enforcer implementations deal with creating the clients that they need. This also drops the need for a provider-like interface, as the enforcer impls know exactly which clients they need, and can create them on the spot

**Comment by danielfritsch** [comment]
Removed ServiceLoaders and the provider interface and impls too. Instead, the initEnforcer function takes no args, allowing the enforcer implementations to pick and initialize whichever admin clients they need.

**Comment by shrijeet-tesla** [comment]
I noticed that at few places the resource is being closed by an entity which didn't create it. In general, the entity that created the resource must close it. 

avoid this,
```
def foo():
   resource = create()
   bar(resource)

def bar():
  resource.do_stuff()
  resource.close()
```

You could make `Enforcer` an `AutoCloseable` and use the try-with-resource paradigm. Your arg-less `initEnforcer` change is fine.

**Comment by danielfritsch** [comment]
update to close the enforcer in same method as it was created

File: kafka_acl_enforcer/src/main/java/com/tesla/data/acl/AclEnforceCommand.java [hunk 7]
```diff
@@ -32,9 +34,15 @@ public AclEnforceCommand(Map<String, Object> cmdConfig, String cluster) {
...
+    this.adminClient = KafkaAdminClient.create(kafkaConfig());
     return new AclEnforcer(configuredEntities(AclBinding.class, "acls", "aclsFile"),
-        new AclService(client), !unsafemode, dryrun);
+        new AclService(adminClient), !unsafemode, dryrun);
   }
-}
 
+  @Override
+  protected void close() {
+    assert this.adminClient != null;
```
Comment (shrijeet-tesla) [inline kafka_acl_enforcer/src/main/java/com/tesla/data/acl/AclEnforceCommand.java:45]: I'd change this to an if condition as opposed to an assert. Also, `assert` is not guaranteed to work, they are disabled by default https://docs.oracle.com/cd/E19683-01/806-7930/6jgp65ikq/index.html

Comment (danielfritsch) [reply kafka_acl_enforcer/src/main/java/com/tesla/data/acl/AclEnforceCommand.java]: updated

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java [hunk 8]
```diff
@@ -19,13 +20,24 @@
...
+    protected Enforcer<ConfiguredTopic> initEnforcer() {
+      this.adminClient = KafkaAdminClient.create(kafkaConfig());
       return new TopicEnforcer(
           new TopicServiceImpl(adminClient, dryrun),
           configuredEntities(ConfiguredTopic.class, "topics", "topicsFile"),
           !unsafemode);
     }
+
+    @Override
+    public void close() {
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java:35]:
```
if (foo != null):
   foo.something()
```
.. would suffice.

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicService.java [hunk 9]
```diff
@@ -51,4 +51,5 @@
    * @param topics a collection of topics
    */
   void delete(List<ConfiguredTopic> topics);
+
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicService.java:54]: unrelated change?

**Review by shrijeet-tesla — APPROVED** [review]

File: kafka_acl_enforcer/src/main/java/com/tesla/data/acl/AclEnforcer.java [hunk 10]
```diff
@@ -34,4 +34,5 @@ protected void delete(List<AclBinding> toDelete) {
       aclService.delete(toDelete);
     }
   }
+
```
Comment (jyates) [inline kafka_acl_enforcer/src/main/java/com/tesla/data/acl/AclEnforcer.java:37]: unrelated?

File: kafka_enforcer_common/src/test/java/com/tesla/data/enforcer/DummyEnforcer.java [hunk 11]
```diff
@@ -31,4 +31,5 @@ protected void create(List<Dummy> toCreate) {
   protected void delete(List<Dummy> toDelete) {
     // no-op
   }
+
```
Comment (jyates) [inline kafka_enforcer_common/src/test/java/com/tesla/data/enforcer/DummyEnforcer.java:34]: unrelated?

**Review by jyates — APPROVED** [review]
lgtm. Mind just addressing the nits and then squashing all the changes down into a commit with an appropriate commit message?
